service/storage_proxy.cc

```diff
@@ -6545,7 +6545,37 @@ db::hints::manager& storage_proxy::hints_manager_for(db::write_type type) {
 }
 
 future<> storage_proxy::truncate_blocking(sstring keyspace, sstring cfname, std::optional<std::chrono::milliseconds> timeout_in_ms) {
-    return remote().send_truncate_blocking(std::move(keyspace), std::move(cfname), timeout_in_ms);
+    locator::effective_replication_map_ptr erm;
+    const auto& ks = _db.local().find_keyspace(keyspace);
+    if (ks.get_replication_strategy().uses_tablets()) {
+        slogger.debug("Running TRUNCATE TABLE (with tablets) on {}.{}", keyspace, cfname);
+        co_await utils::get_local_injector().inject("truncate_with_tablets_wait_before", [] (auto& handler) {
+            slogger.debug("truncate_with_tablets_wait_before: start");
+            return handler.wait_for_message(db::timeout_clock::now() + std::chrono::minutes(2));
+        });
+
+        // Holding the erm pointer will pause load balancing until the pointer is released
+        const auto wait_started = db::timeout_clock::now();
+        erm = co_await _ss.local().get_erm_when_table_has_no_migrations(keyspace, cfname);
+        slogger.debug("Holding ERM pointer for TRUNCATE TABLE (with tablets) on {}.{}", keyspace, cfname);
+        const auto wait_duration_ms = std::chrono::duration_cast<std::chrono::milliseconds>(db::timeout_clock::now() - wait_started);
+        if (timeout_in_ms && wait_duration_ms >= timeout_in_ms) {
+            throw std::runtime_error(format("Timeout for TRUNCATE TABLE on {}.{}", keyspace, cfname));
```

Comment:
We should throw timed_out_error which is a standard error for timeouts, which upper layers may choose to handle.

Reply:
I have changed the implementation to throw `timed_out_error`